hugr-passes/src/linearize_array.rs

```diff
@@ -0,0 +1,418 @@
+//! Provides [LinearizeArrayPass] which turns 'value_array`s into regular linear `array`s.
+
+use hugr_core::{
+    extension::{
+        prelude::Noop,
+        simple_op::{HasConcrete, MakeRegisteredOp},
+    },
+    hugr::hugrmut::HugrMut,
+    ops::NamedOp,
+    std_extensions::collections::{
+        array::{
+            array_type_def, array_type_parametric, Array, ArrayKind, ArrayOpDef, ArrayRepeatDef,
+            ArrayScanDef, ArrayValue, ARRAY_REPEAT_OP_ID, ARRAY_SCAN_OP_ID,
+        },
+        value_array::{self, VArrayFromArrayDef, VArrayToArrayDef, VArrayValue, ValueArray},
+    },
+    types::Transformable,
+    Node,
+};
+use itertools::Itertools;
+use strum::IntoEnumIterator;
+
+use crate::{
+    replace_types::{
+        handlers::copy_discard_array, DelegatingLinearizer, NodeTemplate, ReplaceTypesError,
+    },
+    ComposablePass, ReplaceTypes,
+};
+
+/// A HUGR -> HUGR pass that turns 'value_array`s into regular linear `array`s.
+///
+/// To construct a [LinearizeArrayPass] use [Default::default].
+#[derive(Clone)]
+pub struct LinearizeArrayPass(ReplaceTypes);
+
+impl Default for LinearizeArrayPass {
+    fn default() -> Self {
+        let mut pass = ReplaceTypes::default();
+        pass.replace_parametrized_type(ValueArray::type_def(), |args| {
+            Some(Array::ty_parametric(args[0].clone(), args[1].clone()).unwrap())
+        });
+        pass.replace_consts_parametrized(ValueArray::type_def(), |v, replacer| {
+            let v: &VArrayValue = v.value().downcast_ref().unwrap();
+            let mut ty = v.get_element_type().clone();
+            let mut contents = v.get_contents().iter().cloned().collect_vec();
+            ty.transform(replacer).unwrap();
+            contents.iter_mut().for_each(|v| {
+                replacer.change_value(v).unwrap();
+            });
+            Ok(Some(ArrayValue::new(ty, contents).into()))
+        });
+        for op_def in ArrayOpDef::iter() {
+            pass.replace_parametrized_op(
+                value_array::EXTENSION.get_op(&op_def.name()).unwrap(),
+                move |args| {
+                    // `get` is only allowed for copyable elements. If the elements
+                    // are now linear, we'd have to replace it with a `set`. But what
+                    // should we put in??? For now, let's just error out and make sure
+                    // we're not emitting `get`s for nested value arrays.
+                    if op_def == ArrayOpDef::get && !args[1].as_type().unwrap().copyable() {
```

Comment:
This check avoids `get`s for all linear types, not just nested value arrays, right? The error message should probably reflect that. It'd also be good to document this limitation in some doc comment on the pass.

Reply:
Validity checking should ensure that the type was copyable before we ran the pass. The only way for the element to become linear is if it used to contain a `value_array` that we turned into a linear `array`.

I'll add it to the doc 👍